### DISCUSSION TITLE
Confusing Parse Error 302

### DISCUSSION THREAD

**Discussion by pes-ethanp** [nlohmann/json#3526] [Q&A] [answered]
Writing an application for parsing JSON in string form coming via TCP.
I'm seeing a JSON parse error 302 as follows:
`what(): [json.exception.type_error.302] type must be number, but is number`

This is the error for multiple different JSON messages. Normally this error makes sense but this one is obviously confusing. 

I tried enabling JSON_DIAGNOSTICS to see where it was popping but this causes a seg fault.

Don't know what other context would help here considering the puzzling error...

**Comment by falbrechtskirchinger** [comment, marked as answer]
There are only two locations where this error message could come from. Are you using `nlohmann::json` or are you customizing the `basic_json` template in any way? Have you defined any custom type conversions? Which version of the library are you using?

Meanwhile, I'll keep staring at the code to see why this might be happening.

**Reply by falbrechtskirchinger** [reply]
From looking at the code, I'm none the wiser as to what might be going on here.

I'd suggest you save one of those problematic JSON messages to a file and remove parts of the message until parsing succeeds. Once you've identified the text causing the error, we can help you further. I assume you can't share the JSON, or you would have done so.

**Reply by pes-ethanp** [reply]
I'm using the nlohmann json.hpp, version 3.10.5, no customization of the basic_json template, no custom type conversions. 

The source of the JSON being sent over TCP is a sensor, although I'm reasonably confident that there's no JSON customization being done on their end either.

More specifically, with a more normal string tokenization done with the newline character as the delimiter, I was having no issues parsing the JSONs. I only started encountering this issue when I started splitting manually with delimiter as the sequence of "}{" brackets, which I was doing because some of the JSON messages being sent over were not being finished with a newline character and were going right into the next message, and therefore were being filtered out by the tokenization and caught in the try catch for parse errors. 

Wondering if this manual splitting - which to roughly summarize I'm doing with strings, substrings, string vectors, and recursion to keep splitting until "}{" isn't found - is causing some pointer havoc and subsequently memory. Which would sort of explain the seg fault from the json diagnostics.

**Reply by falbrechtskirchinger** [reply]
> Wondering if this manual splitting - which to roughly summarize I'm doing with strings, substrings, string vectors, and recursion to keep splitting until "}{" isn't found - is causing some pointer havoc and subsequently memory. Which would sort of explain the seg fault from the json diagnostics.

Worth looking into. I can't explain it any other way at the moment.

**Reply by pes-ethanp** [reply]
The sensor documentation is public and no customization is involved, so I can share. The two messages below:
`{
  "time": 106.3935775756836,
  "vx": -3.713480691658333e-05,
  "vy": 5.703703573090024e-05,
  "vz": 2.4990416932269e-05,
  "fom": 0.00016016385052353144,
  "covariance": [
    [
      2.4471841442164077e-08,
      -3.3937477272871774e-09,
      -1.6659699175747278e-09
    ],
    [
      -3.3937477272871774e-09,
      1.4654466085062268e-08,
      4.0409570134514183e-10
    ],
    [
      -1.6659699175747278e-09,
      4.0409570134514183e-10,
      1.5971971523143225e-09
    ]
  ],
  "altitude": 0.4949815273284912,
  "transducers": [
    {
      "id": 0,
      "velocity": 0.00010825289791682735,
      "distance": 0.5568000078201294,
      "rssi": -30.494251251220703,
      "nsd": -88.73271179199219,
      "beam_valid": true
    },
    {
      "id": 1,
      "velocity": -1.4719001228513662e-05,
      "distance": 0.5663999915122986,
      "rssi": -31.095735549926758,
      "nsd": -89.5116958618164,
      "beam_valid": true
    },
    {
      "id": 2,
      "velocity": 2.7863150535267778e-05,
      "distance": 0.537600040435791,
      "rssi": -27.180519104003906,
      "nsd": -96.98075103759766,
      "beam_valid": true
    },
    {
      "id": 3,
      "velocity": 1.9419496311456896e-05,
      "distance": 0.5472000241279602,
      "rssi": -28.006759643554688,
      "nsd": -88.32147216796875,
      "beam_valid": true
    }
  ],
  "velocity_valid": true,
  "status": 0,
  "format": "json_v3",
  "type": "velocity",
  "time_of_validity": 1638191471563017,
  "time_of_transmission": 1638191471752336
}`
`{
  "ts": 49056.809,
  "x": 12.43563613697886467,
  "y": 64.617631152402609587,
  "z": 1.767641898933798075,
  "std": 0.001959984190762043,
  "roll": 0.6173566579818726,
  "pitch": 0.6173566579818726,
  "yaw": 0.6173566579818726,
  "type": "position_local",
  "status": 0,
  "format": "json_v3"
}
`

**Reply by falbrechtskirchinger** [reply]
There is a PR pending that might help you. The dev has been MIA, unfortunately.

@nlohmann Do you want me to look into that and maybe finish the PR by Sunday in time for 3.11.0?

**Reply by pes-ethanp** [reply]
Okay, I'll keep an eye out. Thank you for your help.

**Reply by gregmarr** [reply]
@pes-ethanp Error 302 is not a parsing error.  It's a data extraction error.  Specifically, this error with "type must be number" can only happen in two places, as @falbrechtskirchinger mentioned, both involving `from_json`:

get_arithmetic_value
https://github.com/nlohmann/json/blob/6b97599a274b9b72caffa1332d5384c9aac27590/include/nlohmann/detail/conversions/from_json.hpp#L84

from_json(const BasicJsonType& j, ArithmeticType& val)
https://github.com/nlohmann/json/blob/6b97599a274b9b72caffa1332d5384c9aac27590/include/nlohmann/detail/conversions/from_json.hpp#L374

Can you verify that you're using the HEAD of the `develop` branch or something remotely close to that?  Do you have a small reproduction scenario that shows what you're doing after the json::parse() that's causing the error?

https://godbolt.org/z/4GzWr6Kdq

**Reply by pes-ethanp** [reply]
I'm using the latest release version of the single json.hpp header, so I'll update the header file to the one in the develop branch and give it a whirl when I get a chance. 

Also based on my debugging it was precisely the call to the json::parse() that was causing this error. It did not get past that.

**Reply by pes-ethanp** [reply]
Updated header file does not change outcome, unfortunately. However, I've been told by the sensor manufacturer that they'll be fixing the newline issue with the JSON message in the near future, so I probably won't be looking into this issue much further. That doesn't mean this extraction type error is addressed, so feel free to close/mark as answered/leave as is as appropriate.

Cheers.

**Reply by nlohmann** [reply]
> @nlohmann Do you want me to look into that and maybe finish the PR by Sunday in time for 3.11.0?

That would be great!

**Reply by nlohmann** [reply]
@pes-ethanp Any news on this?

**Reply by falbrechtskirchinger** [reply]
FYI, PR #3548 demonstrates another way to parse multiple consecutive JSON values using iterators — no need to insert new lines into the input.

**Reply by pes-ethanp** [reply]
The sensor manufacturer published a firmware update which fixed the issue of certain JSON messages not ending with a newline character, so a conventional splitting worked just fine. Haven't had the bandwidth to go back and investigate the original parsing error any further.

**Comment by doomroep5** [comment]
> Writing an application for parsing JSON in string form coming via TCP. I'm seeing a JSON parse error 302 as follows: `what(): [json.exception.type_error.302] type must be number, but is number`
> 
> This is the error for multiple different JSON messages. Normally this error makes sense but this one is obviously confusing.
> 
> I tried enabling JSON_DIAGNOSTICS to see where it was popping but this causes a seg fault.
> 
> Don't know what other context would help here considering the puzzling error...
